docs/apache-airflow/concepts/scheduler.rst

```diff
@@ -138,18 +141,172 @@ The following databases are fully supported and provide an "optimal" experience:
+
+What resources might limit Scheduler's performance
+""""""""""""""""""""""""""""""""""""""""""""""""""
+
+There are several areas of resource usage that you should pay attention to:
+
+* FileSystem performance. Airflow Scheduler relies heavily on parsing (sometimes a lot) of Python
+  files, which are often located on a shared filesystem. Airflow Scheduler continuously reads and
+  re-parses those files. The same files have to be made available to workers, so often they are
+  stored in a distributed filesystem. You can use various filesystems for that purpose (NFS, CIFS, EFS,
+  GCS fuse, Azure File System are good examples). There are various parameters you can control for those
+  filesystems and fine-tune their performance, but this is beyond the scope of this document. You should
+  observe statistics and usage of your filesystem to determine if problems come from the filesystem
+  performance. For example there are anecdotal evidences that increasing IOPS (and paying more) for the
+  EFS performance, dramatically improves stability and speed of parsing Airflow DAGs when EFS is used.
+* Another solution to FileSystem performance, if it becomes your bottleneck, is to turn to alternative
+  mechanisms of distributing your DAGs. Embedding DAGs in your image and GitSync distribution have both
+  the property that the files are available locally for Scheduler and it does not have to use a
+  distributed filesystem to read the files, the files are available locally for the Scheduler and it is
+  usually as fast as it can be, especially if your machines use fast SSD disks for local storage. Those
+  distribution mechanisms have other characteristics that might make them not the best choice for you,
+  but if your problems with performance come from distributed filesystem performance, they might be the
+  best approach to follow.
+* Database connections and Database usage might become a problem as you want to increase performance and
+  process more things in parallel. Airflow is known from being "database-connection hungry" - the more DAGs
+  you have and the more you want to process in parallel, the more database connections will be opened.
+  This is generally not a problem for MySQL as its model of handling connections is thread-based, but this
+  might be a problem for Postgres, where connection handling is process-based. It is a general consensus
+  that if you have even medium size Postgres-based Airflow installation, the best solution is to use
+  `PGBouncer <https://www.pgbouncer.org/>`_ as a proxy to your database. The :doc:`helm-chart:index`
+  supports PGBouncer out-of-the-box. For MsSQL we have not yet worked out the best practices as support
+  for MsSQL is still experimental.
+* CPU usage is most important for FileProcessors - those are the processes that parse and execute
+  Python DAG files. Since Schedulers triggers such parsing continuously, when you have a lot complex DAGs,
```

Comment:
What is a complex DAG? do we mean just high number of tasks?

Reply:
I think there are multiple things that might be seen as "complex" , For example "one that takes a lot of time to parse", generates "plenty of dags", "generates plenty of tasks", "generates plenty of tasks interconnection". I have no (and I think no-one else has) a ready to give recipe "this is too complex" or "this is simple enough". 

Each user can write their dags in their own ways. We can determine some bad practices (and we did already - with external DB usage for example). Python is so versatile that it's difficult to say "this is complex" this is "simple". However it's rather easy to say "this is less complex than it's been before" when you simplify it" and I think users should be gently guided that they should understand what they do when they write Python code and work out their own "ways" of developing DAGs which lead to less complexity.

But this is a good point to point people to some ways of at least assessing it - we have the right chapter in `best practices` about testing dag - for example "check tha your DAG imports <fast>" and try to speed it up, if you perceive this being too long (again we cannot tell what "fast" is  - for some users this might be ms, for some several seconds is still ok. But every one of those users can be told `if you want to improve performance - made DAG loading "faster", and here is the tool to test it`. This is precisely what I am going to do.

I will describe the above. I think users should be made aware that there is no ready-to-use-recipe for simple vs. complex and they should simply strive to improve their DAG in terms of loading time, number of Tasks and interconnections between them.